doc/arch/adr-006-geoprocessing-multioperation.md

```diff
@@ -0,0 +1,167 @@
+
+Furthermore, it is our understanding that fetching tiles from S3 is the largest
+portion of a single request's execution time. Thus, removing its duplication is
+the greatest speedup we could achieve in a single step.
+
+## Proposed Solution
+
+To optimize and reuse tile fetches, the geoprocessing service needs to be aware
+that they will be reused. Thus, rather than performing an operation for a
+single shape, the geoprocessing service should take a set of shapes. In this
+set, one should be the parent shape, and all the rest should be the component
+shapes. The parent shape will be used to fetch tiles from S3 and store them in
+memory. Then, the operation will be performed on every component shape,
+choosing the intersecting subset of tiles already in memory. Instead of the
+output being a single set of results, it will be a list of sets of results, in
+the same order as the incoming component shapes.
+
+The above can be done either to the existing operations, and adapting their
+output in MMW to account for the new list return format. Previous operations
+can still be supported with just a singleton list as the output. Or the above
+can be implemented as new, "multishape" versions of the same operations.
+
+Furthermore, since a number of the geoprocessing calls invoked by MapShed query
+the same rasters, e.g. [n][github 9][l][github 10][c][github 11][d][github 12],
+including them in them in the same geoprocessing call would be a speed up as
+well. This would require not only sending a set of shapes, but a set of
+operations as well, each with an `operationType` and a list of `rasters`. The
+output should be a list of lists of sets of results, where the first list
+matches the list of component shapes, and the second the list of operations.
+
+Since both the input and output of this operation can be quite different than
+the current implementation, it should be added as a new operation type rather
+than a modification to an existing one. Subbasin modeling is just a wrapper
+around MapShed, as it decomposes a parent shape into child shapes and runs
+MapShed on each of them, before aggregating the results together. In order to
+fully take advantage of the optimization, MapShed itself must be updated to use
+the new operation. The new operation should also support submitting a single
+shape instead of a list of shapes, i.e. the use case for a single HUC-12. In
+this case the output should be a singleton list of list of results. This will
+allow regular MapShed to use the same operation, rather than have a competing
+implementation for Subbasin modeling.
+
+## Decision
+
+We will add a new operation to the geoprocessing service that takes a list of
+shapes and a list of operations. The first shape in the list is assumed to be
+the parent shape that contains all the other child shapes. Using the parent
+shape we pull tiles corresponding to every raster in the list of operations,
+and collect them in a [`CompositeTile`][github 13] per raster.
```

Comment:
Let me call out our discussion here:

`CompositeTile` essentially provides `Seq[(SpatialKey, Tile)] => Tile`. This would make it easier to express operations over the area because you could treat the query result as a single raster without worrying about tile boundaries and the associated code of iterating over tiles.

However as @rajadain pointed out the current code is already implemented as [operation over `Seq(SpatialKey, Tile)](https://github.com/WikiWatershed/mmw-geoprocessing/blob/c212f29e0b090dd9624b8f637e59fdf6cbbf4f4c/api/src/main/scala/Geoprocessing.scala#L253-L262) so its unlikely that this refactor would be a benefit in this task.

Additionally the task [parallelizes](https://github.com/WikiWatershed/mmw-geoprocessing/blob/c212f29e0b090dd9624b8f637e59fdf6cbbf4f4c/api/src/main/scala/Geoprocessing.scala#L252) over each individual tile. So converting to `CompositeTile` would remove that level of parallelism. This is something to consider but it likely won't have noticeable effect on performance because the proposed changes could/should parallelize over operations per query geometry.

TLDR: Introducing `CompositeTile` usage would simplify the code and likely offset the complexity introduced by handling multiple areas and operations per request but isn't strictly necessary.

Reply:
Alright I'll add a note that makes `ComposeTile` refactor optional.